google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java

```diff
@@ -442,22 +591,163 @@
+          }
+        });
+  }
+
+  /**
+   * Schedules the provided write operation and runs the user success callback when the write result
+   * is obtained.
+   */
+  private ApiFuture<WriteResult> executeWrite(
+      final DocumentReference documentReference,
+      final OperationType operationType,
+      final BulkWriterOperationCallback operationCallback) {
+    final SettableApiFuture<Void> operationCompletedFuture = SettableApiFuture.create();
+
+    ApiFuture<WriteResult> writeResultApiFuture =
+        ApiFutures.transformAsync(
+            enqueueWrite(documentReference, operationType, operationCallback),
+            new ApiAsyncFunction<WriteResult, WriteResult>() {
+              public ApiFuture<WriteResult> apply(final WriteResult result)
+                  throws ExecutionException, InterruptedException {
+
+                invokeUserSuccessCallback(documentReference, result).get();
+                return ApiFutures.immediateFuture(result);
+              }
+            },
+            MoreExecutors.directExecutor());
+
+    writeResultApiFuture.addListener(
+        new Runnable() {
+          public void run() {
+            pendingOperations.remove(operationCompletedFuture);
+            operationCompletedFuture.set(null);
+          }
+        },
+        MoreExecutors.directExecutor());
+
+    pendingOperations.add(operationCompletedFuture);
+    return writeResultApiFuture;
+  }
+
+  /** Used to schedule the write enqueue logic onto the bulkWriterExecutor. */
+  private ApiFuture<WriteResult> enqueueWrite(
+      final DocumentReference documentReference,
+      final OperationType operationType,
+      final BulkWriterOperationCallback operationCallback) {
+    return ApiFutures.transformAsync(
+        ApiFutures.immediateFuture(null),
+        new ApiAsyncFunction<Object, WriteResult>() {
+          public ApiFuture<WriteResult> apply(Object o) throws Exception {
+            return enqueueWriteHelper(documentReference, operationType, operationCallback, 0);
+          }
+        },
+        bulkWriterExecutor);
+  }
+
+  /**
+   * Adds the write to the appropriate batch queue and performs the bulkCommit. This helper method
+   * also catches failures and applies the user error callback and retrying if necessary.
+   */
+  private ApiFuture<WriteResult> enqueueWriteHelper(
+      final DocumentReference documentReference,
+      final OperationType operationType,
+      final BulkWriterOperationCallback operationCallback,
+      final int failedAttempts) {
+    List<BulkCommitBatch> operationBatchQueue = failedAttempts > 0 ? retryBatchQueue : batchQueue;
+    BulkCommitBatch bulkCommitBatch = getEligibleBatch(documentReference, operationBatchQueue);
+
+    // Send ready batches if this is the first attempt. Subsequent retry batches are scheduled
+    // after the initial batch returns.
+    if (failedAttempts == 0) {
+      sendReadyBatches(operationBatchQueue);
```

Comment:
Can we move this to `enqueueWrite`? I think this would allow you to get rid of this if-check.